index out of bound panic

a panic error was reported, here is the full reconcile cycle log
```
{"level":"info","ts":1586453389.1349304,"logger":"egressipam-controller","msg":"Reconciling EgressIPAM","Request.Namespace":"","Request.Name":"egressipam-aws"}
{"level":"debug","ts":1586453389.1355455,"logger":"egressipam-controller","msg":"currently assigned ","IPs by CIDR":{"10.183.0.0/24":["10.183.0.9","10.183.0.6","10.183.0.13","10.183.0.16","10.183.0.15","10.183.0.5","10.183.0.10","10.183.0.14","10.183.0.4","10.183.0.7","10.183.0.11","10.183.0.12"],"10.183.1.0/24":["10.183.1.7","10.183.1.4","10.183.1.11","10.183.1.14","10.183.1.13","10.183.1.5","10.183.1.8","10.183.1.12","10.183.1.74","10.183.1.6","10.183.1.9","10.183.1.10"],"10.183.2.0/24":["10.183.2.7","10.183.2.5","10.183.2.11","10.183.2.14","10.183.2.13","10.183.2.180","10.183.2.8","10.183.2.12","10.183.2.4","10.183.2.6","10.183.2.9","10.183.2.10"]}}
{"level":"debug","ts":1586453389.135592,"logger":"egressipam-controller","msg":"adding always excluded network IPs ","IPs by CIDR":{"10.183.0.0/24":["10.183.0.9","10.183.0.6","10.183.0.13","10.183.0.16","10.183.0.15","10.183.0.5","10.183.0.10","10.183.0.14","10.183.0.4","10.183.0.7","10.183.0.11","10.183.0.12","10.183.0.0","10.183.0.255"],"10.183.1.0/24":["10.183.1.7","10.183.1.4","10.183.1.11","10.183.1.14","10.183.1.13","10.183.1.5","10.183.1.8","10.183.1.12","10.183.1.74","10.183.1.6","10.183.1.9","10.183.1.10","10.183.1.0","10.183.1.255"],"10.183.2.0/24":["10.183.2.7","10.183.2.5","10.183.2.11","10.183.2.14","10.183.2.13","10.183.2.180","10.183.2.8","10.183.2.12","10.183.2.4","10.183.2.6","10.183.2.9","10.183.2.10","10.183.2.0","10.183.2.255"]}}
{"level":"debug","ts":1586453389.135617,"logger":"egressipam-controller","msg":"adding reserved IPs ","IPs by CIDR":{"10.183.0.0/24":["10.183.0.9","10.183.0.6","10.183.0.13","10.183.0.16","10.183.0.15","10.183.0.5","10.183.0.10","10.183.0.14","10.183.0.4","10.183.0.7","10.183.0.11","10.183.0.12","10.183.0.0","10.183.0.255","10.183.0.8","10.183.0.37","10.183.0.142","10.183.0.126","10.183.0.119","10.183.0.103"],"10.183.1.0/24":["10.183.1.7","10.183.1.4","10.183.1.11","10.183.1.14","10.183.1.13","10.183.1.5","10.183.1.8","10.183.1.12","10.183.1.74","10.183.1.6","10.183.1.9","10.183.1.10","10.183.1.0","10.183.1.255","10.183.1.102","10.183.1.144","10.183.1.149","10.183.1.15","10.183.1.225","10.183.1.54"],"10.183.2.0/24":["10.183.2.7","10.183.2.5","10.183.2.11","10.183.2.14","10.183.2.13","10.183.2.180","10.183.2.8","10.183.2.12","10.183.2.4","10.183.2.6","10.183.2.9","10.183.2.10","10.183.2.0","10.183.2.255","10.183.2.141","10.183.2.254","10.183.2.76"]}}
{"level":"debug","ts":1586453389.1358423,"logger":"egressipam-controller","msg":"adding nodes IPs (if in the same CIDR) ","IPs by CIDR":{"10.183.0.0/24":["10.183.0.9","10.183.0.6","10.183.0.13","10.183.0.16","10.183.0.15","10.183.0.5","10.183.0.10","10.183.0.14","10.183.0.4","10.183.0.7","10.183.0.11","10.183.0.12","10.183.0.0","10.183.0.255","10.183.0.8","10.183.0.37","10.183.0.142","10.183.0.126","10.183.0.119","10.183.0.103","10.183.0.129","10.183.0.210","10.183.0.68"],"10.183.1.0/24":["10.183.1.7","10.183.1.4","10.183.1.11","10.183.1.14","10.183.1.13","10.183.1.5","10.183.1.8","10.183.1.12","10.183.1.74","10.183.1.6","10.183.1.9","10.183.1.10","10.183.1.0","10.183.1.255","10.183.1.102","10.183.1.144","10.183.1.149","10.183.1.15","10.183.1.225","10.183.1.54","10.183.1.246","10.183.1.79","10.183.1.120"],"10.183.2.0/24":["10.183.2.7","10.183.2.5","10.183.2.11","10.183.2.14","10.183.2.13","10.183.2.180","10.183.2.8","10.183.2.12","10.183.2.4","10.183.2.6","10.183.2.9","10.183.2.10","10.183.2.0","10.183.2.255","10.183.2.141","10.183.2.254","10.183.2.76","10.183.2.240","10.183.2.161","10.183.2.144"]}}
{"level":"debug","ts":1586453389.1358724,"logger":"egressipam-controller","msg":"adding cloud infrastructure reserved IPs ","IPs by CIDR":{"10.183.0.0/24":["10.183.0.9","10.183.0.6","10.183.0.13","10.183.0.16","10.183.0.15","10.183.0.5","10.183.0.10","10.183.0.14","10.183.0.4","10.183.0.7","10.183.0.11","10.183.0.12","10.183.0.0","10.183.0.255","10.183.0.8","10.183.0.37","10.183.0.142","10.183.0.126","10.183.0.119","10.183.0.103","10.183.0.129","10.183.0.210","10.183.0.68","10.183.0.1","10.183.0.2","10.183.0.3"],"10.183.1.0/24":["10.183.1.7","10.183.1.4","10.183.1.11","10.183.1.14","10.183.1.13","10.183.1.5","10.183.1.8","10.183.1.12","10.183.1.74","10.183.1.6","10.183.1.9","10.183.1.10","10.183.1.0","10.183.1.255","10.183.1.102","10.183.1.144","10.183.1.149","10.183.1.15","10.183.1.225","10.183.1.54","10.183.1.246","10.183.1.79","10.183.1.120","10.183.1.1","10.183.1.2","10.183.1.3"],"10.183.2.0/24":["10.183.2.7","10.183.2.5","10.183.2.11","10.183.2.14","10.183.2.13","10.183.2.180","10.183.2.8","10.183.2.12","10.183.2.4","10.183.2.6","10.183.2.9","10.183.2.10","10.183.2.0","10.183.2.255","10.183.2.141","10.183.2.254","10.183.2.76","10.183.2.240","10.183.2.161","10.183.2.144","10.183.2.1","10.183.2.2","10.183.2.3"]}}
{"level":"debug","ts":1586453389.135928,"logger":"egressipam-controller","msg":"sorted reserved IPs ","IPs by CIDR":{"10.183.0.0/24":["10.183.0.0","10.183.0.1","10.183.0.2","10.183.0.3","10.183.0.4","10.183.0.5","10.183.0.6","10.183.0.7","10.183.0.8","10.183.0.9","10.183.0.10","10.183.0.11","10.183.0.12","10.183.0.13","10.183.0.14","10.183.0.15","10.183.0.16","10.183.0.37","10.183.0.68","10.183.0.103","10.183.0.119","10.183.0.126","10.183.0.129","10.183.0.142","10.183.0.210","10.183.0.255"],"10.183.1.0/24":["10.183.1.0","10.183.1.1","10.183.1.2","10.183.1.3","10.183.1.4","10.183.1.5","10.183.1.6","10.183.1.7","10.183.1.8","10.183.1.9","10.183.1.10","10.183.1.11","10.183.1.12","10.183.1.13","10.183.1.14","10.183.1.15","10.183.1.54","10.183.1.74","10.183.1.79","10.183.1.102","10.183.1.120","10.183.1.144","10.183.1.149","10.183.1.225","10.183.1.246","10.183.1.255"],"10.183.2.0/24":["10.183.2.0","10.183.2.1","10.183.2.2","10.183.2.3","10.183.2.4","10.183.2.5","10.183.2.6","10.183.2.7","10.183.2.8","10.183.2.9","10.183.2.10","10.183.2.11","10.183.2.12","10.183.2.13","10.183.2.14","10.183.2.76","10.183.2.141","10.183.2.144","10.183.2.161","10.183.2.180","10.183.2.240","10.183.2.254","10.183.2.255"]}}
{"level":"debug","ts":1586453389.13595,"logger":"egressipam-controller","msg":"Assigning new IPs from","IPs by CIDR":{"10.183.0.0/24":["10.183.0.0","10.183.0.1","10.183.0.2","10.183.0.3","10.183.0.4","10.183.0.5","10.183.0.6","10.183.0.7","10.183.0.8","10.183.0.9","10.183.0.10","10.183.0.11","10.183.0.12","10.183.0.13","10.183.0.14","10.183.0.15","10.183.0.16","10.183.0.37","10.183.0.68","10.183.0.103","10.183.0.119","10.183.0.126","10.183.0.129","10.183.0.142","10.183.0.210","10.183.0.255"],"10.183.1.0/24":["10.183.1.0","10.183.1.1","10.183.1.2","10.183.1.3","10.183.1.4","10.183.1.5","10.183.1.6","10.183.1.7","10.183.1.8","10.183.1.9","10.183.1.10","10.183.1.11","10.183.1.12","10.183.1.13","10.183.1.14","10.183.1.15","10.183.1.54","10.183.1.74","10.183.1.79","10.183.1.102","10.183.1.120","10.183.1.144","10.183.1.149","10.183.1.225","10.183.1.246","10.183.1.255"],"10.183.2.0/24":["10.183.2.0","10.183.2.1","10.183.2.2","10.183.2.3","10.183.2.4","10.183.2.5","10.183.2.6","10.183.2.7","10.183.2.8","10.183.2.9","10.183.2.10","10.183.2.11","10.183.2.12","10.183.2.13","10.183.2.14","10.183.2.76","10.183.2.141","10.183.2.144","10.183.2.161","10.183.2.180","10.183.2.240","10.183.2.254","10.183.2.255"]}}
{"level":"debug","ts":1586453389.1359735,"logger":"egressipam-controller","msg":"Assigning new IP from","CIDR":"10.183.0.0/24","with already assigned IPs":["10.183.0.0","10.183.0.1","10.183.0.2","10.183.0.3","10.183.0.4","10.183.0.5","10.183.0.6","10.183.0.7","10.183.0.8","10.183.0.9","10.183.0.10","10.183.0.11","10.183.0.12","10.183.0.13","10.183.0.14","10.183.0.15","10.183.0.16","10.183.0.37","10.183.0.68","10.183.0.103","10.183.0.119","10.183.0.126","10.183.0.129","10.183.0.142","10.183.0.210","10.183.0.255"]}
{"level":"debug","ts":1586453389.1359885,"logger":"egressipam-controller","msg":"Assigned ","IP":"10.183.0.17","new assigned IPs":["10.183.0.0","10.183.0.1","10.183.0.2","10.183.0.3","10.183.0.4","10.183.0.5","10.183.0.6","10.183.0.7","10.183.0.8","10.183.0.9","10.183.0.10","10.183.0.11","10.183.0.12","10.183.0.13","10.183.0.14","10.183.0.15","10.183.0.16","10.183.0.17","10.183.0.37","10.183.0.68","10.183.0.103","10.183.0.119","10.183.0.126","10.183.0.129","10.183.0.142","10.183.0.210","10.183.0.255"]}
{"level":"debug","ts":1586453389.136002,"logger":"egressipam-controller","msg":"Assigning new IP from","CIDR":"10.183.1.0/24","with already assigned IPs":["10.183.1.0","10.183.1.1","10.183.1.2","10.183.1.3","10.183.1.4","10.183.1.5","10.183.1.6","10.183.1.7","10.183.1.8","10.183.1.9","10.183.1.10","10.183.1.11","10.183.1.12","10.183.1.13","10.183.1.14","10.183.1.15","10.183.1.54","10.183.1.74","10.183.1.79","10.183.1.102","10.183.1.120","10.183.1.144","10.183.1.149","10.183.1.225","10.183.1.246","10.183.1.255"]}
{"level":"debug","ts":1586453389.1360135,"logger":"egressipam-controller","msg":"Assigned ","IP":"10.183.1.16","new assigned IPs":["10.183.1.0","10.183.1.1","10.183.1.2","10.183.1.3","10.183.1.4","10.183.1.5","10.183.1.6","10.183.1.7","10.183.1.8","10.183.1.9","10.183.1.10","10.183.1.11","10.183.1.12","10.183.1.13","10.183.1.14","10.183.1.15","10.183.1.16","10.183.1.54","10.183.1.74","10.183.1.79","10.183.1.102","10.183.1.120","10.183.1.144","10.183.1.149","10.183.1.225","10.183.1.246","10.183.1.255"]}
{"level":"debug","ts":1586453389.1360235,"logger":"egressipam-controller","msg":"Assigning new IP from","CIDR":"10.183.2.0/24","with already assigned IPs":["10.183.2.0","10.183.2.1","10.183.2.2","10.183.2.3","10.183.2.4","10.183.2.5","10.183.2.6","10.183.2.7","10.183.2.8","10.183.2.9","10.183.2.10","10.183.2.11","10.183.2.12","10.183.2.13","10.183.2.14","10.183.2.76","10.183.2.141","10.183.2.144","10.183.2.161","10.183.2.180","10.183.2.240","10.183.2.254","10.183.2.255"]}
{"level":"debug","ts":1586453389.1360333,"logger":"egressipam-controller","msg":"Assigned ","IP":"10.183.2.15","new assigned IPs":["10.183.2.0","10.183.2.1","10.183.2.2","10.183.2.3","10.183.2.4","10.183.2.5","10.183.2.6","10.183.2.7","10.183.2.8","10.183.2.9","10.183.2.10","10.183.2.11","10.183.2.12","10.183.2.13","10.183.2.14","10.183.2.15","10.183.2.76","10.183.2.141","10.183.2.144","10.183.2.161","10.183.2.180","10.183.2.240","10.183.2.254","10.183.2.255"]}
{"level":"debug","ts":1586453389.1360428,"logger":"egressipam-controller","msg":"Assigned","new IPs from":["10.183.0.17","10.183.1.16","10.183.2.15"]," new IPs by CIDR":{"10.183.0.0/24":["10.183.0.0","10.183.0.1","10.183.0.2","10.183.0.3","10.183.0.4","10.183.0.5","10.183.0.6","10.183.0.7","10.183.0.8","10.183.0.9","10.183.0.10","10.183.0.11","10.183.0.12","10.183.0.13","10.183.0.14","10.183.0.15","10.183.0.16","10.183.0.17","10.183.0.37","10.183.0.68","10.183.0.103","10.183.0.119","10.183.0.126","10.183.0.129","10.183.0.142","10.183.0.210","10.183.0.255"],"10.183.1.0/24":["10.183.1.0","10.183.1.1","10.183.1.2","10.183.1.3","10.183.1.4","10.183.1.5","10.183.1.6","10.183.1.7","10.183.1.8","10.183.1.9","10.183.1.10","10.183.1.11","10.183.1.12","10.183.1.13","10.183.1.14","10.183.1.15","10.183.1.16","10.183.1.54","10.183.1.74","10.183.1.79","10.183.1.102","10.183.1.120","10.183.1.144","10.183.1.149","10.183.1.225","10.183.1.246","10.183.1.255"],"10.183.2.0/24":["10.183.2.0","10.183.2.1","10.183.2.2","10.183.2.3","10.183.2.4","10.183.2.5","10.183.2.6","10.183.2.7","10.183.2.8","10.183.2.9","10.183.2.10","10.183.2.11","10.183.2.12","10.183.2.13","10.183.2.14","10.183.2.15","10.183.2.76","10.183.2.141","10.183.2.144","10.183.2.161","10.183.2.180","10.183.2.240","10.183.2.254","10.183.2.255"]}}
{"level":"info","ts":1586453389.1362329,"logger":"egressipam-controller","msg":"ips assigned to","namespace":"ics-egressip-test11","ips":["10.183.0.17","10.183.1.16","10.183.2.15"]}
{"level":"debug","ts":1586453389.142604,"logger":"egressipam-controller","msg":"Assigning new IPs from","IPs by CIDR":{"10.183.0.0/24":["10.183.0.0","10.183.0.1","10.183.0.2","10.183.0.3","10.183.0.4","10.183.0.5","10.183.0.6","10.183.0.7","10.183.0.8","10.183.0.9","10.183.0.10","10.183.0.11","10.183.0.12","10.183.0.13","10.183.0.14","10.183.0.15","10.183.0.16","10.183.0.17","10.183.0.37","10.183.0.68","10.183.0.103","10.183.0.119","10.183.0.126","10.183.0.129","10.183.0.142","10.183.0.210","10.183.0.255"],"10.183.1.0/24":["10.183.1.0","10.183.1.1","10.183.1.2","10.183.1.3","10.183.1.4","10.183.1.5","10.183.1.6","10.183.1.7","10.183.1.8","10.183.1.9","10.183.1.10","10.183.1.11","10.183.1.12","10.183.1.13","10.183.1.14","10.183.1.15","10.183.1.16","10.183.1.54","10.183.1.74","10.183.1.79","10.183.1.102","10.183.1.120","10.183.1.144","10.183.1.149","10.183.1.225","10.183.1.246","10.183.1.255"],"10.183.2.0/24":["10.183.2.0","10.183.2.1","10.183.2.2","10.183.2.3","10.183.2.4","10.183.2.5","10.183.2.6","10.183.2.7","10.183.2.8","10.183.2.9","10.183.2.10","10.183.2.11","10.183.2.12","10.183.2.13","10.183.2.14","10.183.2.15","10.183.2.76","10.183.2.141","10.183.2.144","10.183.2.161","10.183.2.180","10.183.2.240","10.183.2.254","10.183.2.255"]}}
{"level":"debug","ts":1586453389.1426632,"logger":"egressipam-controller","msg":"Assigning new IP from","CIDR":"10.183.0.0/24","with already assigned IPs":["10.183.0.0","10.183.0.1","10.183.0.2","10.183.0.3","10.183.0.4","10.183.0.5","10.183.0.6","10.183.0.7","10.183.0.8","10.183.0.9","10.183.0.10","10.183.0.11","10.183.0.12","10.183.0.13","10.183.0.14","10.183.0.15","10.183.0.16","10.183.0.17","10.183.0.37","10.183.0.68","10.183.0.103","10.183.0.119","10.183.0.126","10.183.0.129","10.183.0.142","10.183.0.210","10.183.0.255"]}
{"level":"debug","ts":1586453389.1426806,"logger":"egressipam-controller","msg":"Assigned ","IP":"10.183.0.18","new assigned IPs":["10.183.0.0","10.183.0.1","10.183.0.2","10.183.0.3","10.183.0.4","10.183.0.5","10.183.0.6","10.183.0.7","10.183.0.8","10.183.0.9","10.183.0.10","10.183.0.11","10.183.0.12","10.183.0.13","10.183.0.14","10.183.0.15","10.183.0.16","10.183.0.17","10.183.0.18","10.183.0.37","10.183.0.68","10.183.0.103","10.183.0.119","10.183.0.126","10.183.0.129","10.183.0.142","10.183.0.210","10.183.0.255"]}
{"level":"debug","ts":1586453389.1427045,"logger":"egressipam-controller","msg":"Assigning new IP from","CIDR":"10.183.1.0/24","with already assigned IPs":["10.183.1.0","10.183.1.1","10.183.1.2","10.183.1.3","10.183.1.4","10.183.1.5","10.183.1.6","10.183.1.7","10.183.1.8","10.183.1.9","10.183.1.10","10.183.1.11","10.183.1.12","10.183.1.13","10.183.1.14","10.183.1.15","10.183.1.16","10.183.1.54","10.183.1.74","10.183.1.79","10.183.1.102","10.183.1.120","10.183.1.144","10.183.1.149","10.183.1.225","10.183.1.246","10.183.1.255"]}
{"level":"debug","ts":1586453389.1427152,"logger":"egressipam-controller","msg":"Assigned ","IP":"10.183.1.17","new assigned IPs":["10.183.1.0","10.183.1.1","10.183.1.2","10.183.1.3","10.183.1.4","10.183.1.5","10.183.1.6","10.183.1.7","10.183.1.8","10.183.1.9","10.183.1.10","10.183.1.11","10.183.1.12","10.183.1.13","10.183.1.14","10.183.1.15","10.183.1.16","10.183.1.17","10.183.1.54","10.183.1.74","10.183.1.79","10.183.1.102","10.183.1.120","10.183.1.144","10.183.1.149","10.183.1.225","10.183.1.246","10.183.1.255"]}
{"level":"debug","ts":1586453389.1427264,"logger":"egressipam-controller","msg":"Assigning new IP from","CIDR":"10.183.2.0/24","with already assigned IPs":["10.183.2.0","10.183.2.1","10.183.2.2","10.183.2.3","10.183.2.4","10.183.2.5","10.183.2.6","10.183.2.7","10.183.2.8","10.183.2.9","10.183.2.10","10.183.2.11","10.183.2.12","10.183.2.13","10.183.2.14","10.183.2.15","10.183.2.76","10.183.2.141","10.183.2.144","10.183.2.161","10.183.2.180","10.183.2.240","10.183.2.254","10.183.2.255"]}
{"level":"debug","ts":1586453389.1427443,"logger":"egressipam-controller","msg":"Assigned ","IP":"10.183.2.16","new assigned IPs":["10.183.2.0","10.183.2.1","10.183.2.2","10.183.2.3","10.183.2.4","10.183.2.5","10.183.2.6","10.183.2.7","10.183.2.8","10.183.2.9","10.183.2.10","10.183.2.11","10.183.2.12","10.183.2.13","10.183.2.14","10.183.2.15","10.183.2.16","10.183.2.76","10.183.2.141","10.183.2.144","10.183.2.161","10.183.2.180","10.183.2.240","10.183.2.254","10.183.2.255"]}
{"level":"debug","ts":1586453389.1427596,"logger":"egressipam-controller","msg":"Assigned","new IPs from":["10.183.0.18","10.183.1.17","10.183.2.16"]," new IPs by CIDR":{"10.183.0.0/24":["10.183.0.0","10.183.0.1","10.183.0.2","10.183.0.3","10.183.0.4","10.183.0.5","10.183.0.6","10.183.0.7","10.183.0.8","10.183.0.9","10.183.0.10","10.183.0.11","10.183.0.12","10.183.0.13","10.183.0.14","10.183.0.15","10.183.0.16","10.183.0.17","10.183.0.18","10.183.0.37","10.183.0.68","10.183.0.103","10.183.0.119","10.183.0.126","10.183.0.129","10.183.0.142","10.183.0.210","10.183.0.255"],"10.183.1.0/24":["10.183.1.0","10.183.1.1","10.183.1.2","10.183.1.3","10.183.1.4","10.183.1.5","10.183.1.6","10.183.1.7","10.183.1.8","10.183.1.9","10.183.1.10","10.183.1.11","10.183.1.12","10.183.1.13","10.183.1.14","10.183.1.15","10.183.1.16","10.183.1.17","10.183.1.54","10.183.1.74","10.183.1.79","10.183.1.102","10.183.1.120","10.183.1.144","10.183.1.149","10.183.1.225","10.183.1.246","10.183.1.255"],"10.183.2.0/24":["10.183.2.0","10.183.2.1","10.183.2.2","10.183.2.3","10.183.2.4","10.183.2.5","10.183.2.6","10.183.2.7","10.183.2.8","10.183.2.9","10.183.2.10","10.183.2.11","10.183.2.12","10.183.2.13","10.183.2.14","10.183.2.15","10.183.2.16","10.183.2.76","10.183.2.141","10.183.2.144","10.183.2.161","10.183.2.180","10.183.2.240","10.183.2.254","10.183.2.255"]}}
{"level":"info","ts":1586453389.1427958,"logger":"egressipam-controller","msg":"ips assigned to","namespace":"ics-egressip-test13","ips":["10.183.0.18","10.183.1.17","10.183.2.16"]}
{"level":"debug","ts":1586453389.150271,"logger":"egressipam-controller","msg":"Assigning new IPs from","IPs by CIDR":{"10.183.0.0/24":["10.183.0.0","10.183.0.1","10.183.0.2","10.183.0.3","10.183.0.4","10.183.0.5","10.183.0.6","10.183.0.7","10.183.0.8","10.183.0.9","10.183.0.10","10.183.0.11","10.183.0.12","10.183.0.13","10.183.0.14","10.183.0.15","10.183.0.16","10.183.0.17","10.183.0.18","10.183.0.37","10.183.0.68","10.183.0.103","10.183.0.119","10.183.0.126","10.183.0.129","10.183.0.142","10.183.0.210","10.183.0.255"],"10.183.1.0/24":["10.183.1.0","10.183.1.1","10.183.1.2","10.183.1.3","10.183.1.4","10.183.1.5","10.183.1.6","10.183.1.7","10.183.1.8","10.183.1.9","10.183.1.10","10.183.1.11","10.183.1.12","10.183.1.13","10.183.1.14","10.183.1.15","10.183.1.16","10.183.1.17","10.183.1.54","10.183.1.74","10.183.1.79","10.183.1.102","10.183.1.120","10.183.1.144","10.183.1.149","10.183.1.225","10.183.1.246","10.183.1.255"],"10.183.2.0/24":["10.183.2.0","10.183.2.1","10.183.2.2","10.183.2.3","10.183.2.4","10.183.2.5","10.183.2.6","10.183.2.7","10.183.2.8","10.183.2.9","10.183.2.10","10.183.2.11","10.183.2.12","10.183.2.13","10.183.2.14","10.183.2.15","10.183.2.16","10.183.2.76","10.183.2.141","10.183.2.144","10.183.2.161","10.183.2.180","10.183.2.240","10.183.2.254","10.183.2.255"]}}
{"level":"debug","ts":1586453389.1503615,"logger":"egressipam-controller","msg":"Assigning new IP from","CIDR":"10.183.0.0/24","with already assigned IPs":["10.183.0.0","10.183.0.1","10.183.0.2","10.183.0.3","10.183.0.4","10.183.0.5","10.183.0.6","10.183.0.7","10.183.0.8","10.183.0.9","10.183.0.10","10.183.0.11","10.183.0.12","10.183.0.13","10.183.0.14","10.183.0.15","10.183.0.16","10.183.0.17","10.183.0.18","10.183.0.37","10.183.0.68","10.183.0.103","10.183.0.119","10.183.0.126","10.183.0.129","10.183.0.142","10.183.0.210","10.183.0.255"]}
{"level":"debug","ts":1586453389.1503801,"logger":"egressipam-controller","msg":"Assigned ","IP":"10.183.0.19","new assigned IPs":["10.183.0.0","10.183.0.1","10.183.0.2","10.183.0.3","10.183.0.4","10.183.0.5","10.183.0.6","10.183.0.7","10.183.0.8","10.183.0.9","10.183.0.10","10.183.0.11","10.183.0.12","10.183.0.13","10.183.0.14","10.183.0.15","10.183.0.16","10.183.0.17","10.183.0.18","10.183.0.19","10.183.0.37","10.183.0.68","10.183.0.103","10.183.0.119","10.183.0.126","10.183.0.129","10.183.0.142","10.183.0.210","10.183.0.255"]}
{"level":"debug","ts":1586453389.1503918,"logger":"egressipam-controller","msg":"Assigning new IP from","CIDR":"10.183.1.0/24","with already assigned IPs":["10.183.1.0","10.183.1.1","10.183.1.2","10.183.1.3","10.183.1.4","10.183.1.5","10.183.1.6","10.183.1.7","10.183.1.8","10.183.1.9","10.183.1.10","10.183.1.11","10.183.1.12","10.183.1.13","10.183.1.14","10.183.1.15","10.183.1.16","10.183.1.17","10.183.1.54","10.183.1.74","10.183.1.79","10.183.1.102","10.183.1.120","10.183.1.144","10.183.1.149","10.183.1.225","10.183.1.246","10.183.1.255"]}
{"level":"debug","ts":1586453389.150403,"logger":"egressipam-controller","msg":"Assigned ","IP":"10.183.1.18","new assigned IPs":["10.183.1.0","10.183.1.1","10.183.1.2","10.183.1.3","10.183.1.4","10.183.1.5","10.183.1.6","10.183.1.7","10.183.1.8","10.183.1.9","10.183.1.10","10.183.1.11","10.183.1.12","10.183.1.13","10.183.1.14","10.183.1.15","10.183.1.16","10.183.1.17","10.183.1.18","10.183.1.54","10.183.1.74","10.183.1.79","10.183.1.102","10.183.1.120","10.183.1.144","10.183.1.149","10.183.1.225","10.183.1.246","10.183.1.255"]}
{"level":"debug","ts":1586453389.150413,"logger":"egressipam-controller","msg":"Assigning new IP from","CIDR":"10.183.2.0/24","with already assigned IPs":["10.183.2.0","10.183.2.1","10.183.2.2","10.183.2.3","10.183.2.4","10.183.2.5","10.183.2.6","10.183.2.7","10.183.2.8","10.183.2.9","10.183.2.10","10.183.2.11","10.183.2.12","10.183.2.13","10.183.2.14","10.183.2.15","10.183.2.16","10.183.2.76","10.183.2.141","10.183.2.144","10.183.2.161","10.183.2.180","10.183.2.240","10.183.2.254","10.183.2.255"]}
{"level":"debug","ts":1586453389.1504238,"logger":"egressipam-controller","msg":"Assigned ","IP":"10.183.2.17","new assigned IPs":["10.183.2.0","10.183.2.1","10.183.2.2","10.183.2.3","10.183.2.4","10.183.2.5","10.183.2.6","10.183.2.7","10.183.2.8","10.183.2.9","10.183.2.10","10.183.2.11","10.183.2.12","10.183.2.13","10.183.2.14","10.183.2.15","10.183.2.16","10.183.2.17","10.183.2.76","10.183.2.141","10.183.2.144","10.183.2.161","10.183.2.180","10.183.2.240","10.183.2.254","10.183.2.255"]}
{"level":"debug","ts":1586453389.150434,"logger":"egressipam-controller","msg":"Assigned","new IPs from":["10.183.0.19","10.183.1.18","10.183.2.17"]," new IPs by CIDR":{"10.183.0.0/24":["10.183.0.0","10.183.0.1","10.183.0.2","10.183.0.3","10.183.0.4","10.183.0.5","10.183.0.6","10.183.0.7","10.183.0.8","10.183.0.9","10.183.0.10","10.183.0.11","10.183.0.12","10.183.0.13","10.183.0.14","10.183.0.15","10.183.0.16","10.183.0.17","10.183.0.18","10.183.0.19","10.183.0.37","10.183.0.68","10.183.0.103","10.183.0.119","10.183.0.126","10.183.0.129","10.183.0.142","10.183.0.210","10.183.0.255"],"10.183.1.0/24":["10.183.1.0","10.183.1.1","10.183.1.2","10.183.1.3","10.183.1.4","10.183.1.5","10.183.1.6","10.183.1.7","10.183.1.8","10.183.1.9","10.183.1.10","10.183.1.11","10.183.1.12","10.183.1.13","10.183.1.14","10.183.1.15","10.183.1.16","10.183.1.17","10.183.1.18","10.183.1.54","10.183.1.74","10.183.1.79","10.183.1.102","10.183.1.120","10.183.1.144","10.183.1.149","10.183.1.225","10.183.1.246","10.183.1.255"],"10.183.2.0/24":["10.183.2.0","10.183.2.1","10.183.2.2","10.183.2.3","10.183.2.4","10.183.2.5","10.183.2.6","10.183.2.7","10.183.2.8","10.183.2.9","10.183.2.10","10.183.2.11","10.183.2.12","10.183.2.13","10.183.2.14","10.183.2.15","10.183.2.16","10.183.2.17","10.183.2.76","10.183.2.141","10.183.2.144","10.183.2.161","10.183.2.180","10.183.2.240","10.183.2.254","10.183.2.255"]}}
{"level":"info","ts":1586453389.1504724,"logger":"egressipam-controller","msg":"ips assigned to","namespace":"ics-egressip-test12","ips":["10.183.0.19","10.183.1.18","10.183.2.17"]}
{"level":"debug","ts":1586453389.8089163,"logger":"egressipam-controller","msg":"","assignedIPsToNodesByCIDR: ":{"10.183.0.0/24":["10.183.0.10","10.183.0.5","10.183.0.12","10.183.0.4","10.183.0.6","10.183.0.13","10.183.0.14","10.183.0.11","10.183.0.15","10.183.0.16","10.183.0.9","10.183.0.7"],"10.183.1.0/24":["10.183.1.9","10.183.1.11","10.183.1.10","10.183.1.12","10.183.1.74","10.183.1.8","10.183.1.13","10.183.1.4","10.183.1.14","10.183.1.6","10.183.1.5","10.183.1.7"],"10.183.2.0/24":["10.183.2.5","10.183.2.9","10.183.2.12","10.183.2.14","10.183.2.4","10.183.2.10","10.183.2.6","10.183.2.13","10.183.2.8","10.183.2.11","10.183.2.7","10.183.2.180"]}}
{"level":"debug","ts":1586453389.8090122,"logger":"egressipam-controller","msg":"","assignedIPsToNamespaceByCIDR: ":{"10.183.0.0/24":["10.183.0.9","10.183.0.6","10.183.0.13","10.183.0.16","10.183.0.15","10.183.0.5","10.183.0.10","10.183.0.14","10.183.0.4","10.183.0.7","10.183.0.11","10.183.0.12","10.183.0.17","10.183.0.18","10.183.0.19"],"10.183.1.0/24":["10.183.1.7","10.183.1.4","10.183.1.11","10.183.1.14","10.183.1.13","10.183.1.5","10.183.1.8","10.183.1.12","10.183.1.74","10.183.1.6","10.183.1.9","10.183.1.10","10.183.1.16","10.183.1.17","10.183.1.18"],"10.183.2.0/24":["10.183.2.7","10.183.2.5","10.183.2.11","10.183.2.14","10.183.2.13","10.183.2.180","10.183.2.8","10.183.2.12","10.183.2.4","10.183.2.6","10.183.2.9","10.183.2.10","10.183.2.15","10.183.2.16","10.183.2.17"]}}
{"level":"debug","ts":1586453389.8090377,"logger":"egressipam-controller","msg":"","toBeAssignedToNodesIPsByCIDR: ":{"10.183.0.0/24":["10.183.0.18","10.183.0.17","10.183.0.19"],"10.183.1.0/24":["10.183.1.17","10.183.1.16","10.183.1.18"],"10.183.2.0/24":["10.183.2.16","10.183.2.17","10.183.2.15"]}}
{"level":"debug","ts":1586453389.809057,"logger":"egressipam-controller","msg":"new","assignedIPsToNodesByCIDR: ":{"10.183.0.0/24":["10.183.0.5","10.183.0.13","10.183.0.14","10.183.0.16","10.183.0.7","10.183.0.10","10.183.0.12","10.183.0.4","10.183.0.6","10.183.0.11","10.183.0.15","10.183.0.9"],"10.183.1.0/24":["10.183.1.5","10.183.1.7","10.183.1.6","10.183.1.11","10.183.1.12","10.183.1.74","10.183.1.10","10.183.1.13","10.183.1.9","10.183.1.8","10.183.1.4","10.183.1.14"],"10.183.2.0/24":["10.183.2.11","10.183.2.12","10.183.2.7","10.183.2.9","10.183.2.10","10.183.2.6","10.183.2.180","10.183.2.5","10.183.2.14","10.183.2.8","10.183.2.4","10.183.2.13"]}}
{"level":"debug","ts":1586453389.8090842,"logger":"egressipam-controller","msg":"new","assignedIPsByNode: ":{"ip-10-183-0-210.ec2.internal":["10.183.0.5","10.183.0.13","10.183.0.14","10.183.0.16","10.183.0.10","10.183.0.12","10.183.0.4","10.183.0.6","10.183.0.11","10.183.0.15"],"ip-10-183-0-68.ec2.internal":["10.183.0.7","10.183.0.9"],"ip-10-183-1-246.ec2.internal":["10.183.1.11","10.183.1.12","10.183.1.74","10.183.1.10","10.183.1.13","10.183.1.9","10.183.1.8","10.183.1.4","10.183.1.14"],"ip-10-183-1-79.ec2.internal":["10.183.1.5","10.183.1.7","10.183.1.6"],"ip-10-183-2-144.ec2.internal":["10.183.2.11","10.183.2.7","10.183.2.10","10.183.2.6","10.183.2.180","10.183.2.8","10.183.2.4","10.183.2.13"],"ip-10-183-2-240.ec2.internal":["10.183.2.12","10.183.2.9","10.183.2.5","10.183.2.14"]}}
E0409 17:29:49.809315       1 runtime.go:78] Observed a panic: runtime.boundsError{x:0, y:0, signed:true, code:0x0} (runtime error: index out of range [0] with length 0)
goroutine 1712 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x197a120, 0xc00234aa40)
	/home/rspazzol/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/home/rspazzol/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/runtime/runtime.go:48 +0x82
panic(0x197a120, 0xc00234aa40)
	/usr/lib/golang/src/runtime/panic.go:679 +0x1b2
github.com/redhat-cop/egressip-ipam-operator/pkg/controller/egressipam.(*ReconcileEgressIPAM).assignIPsToNodes(0xc000982900, 0xc0015352c0, 0xc00013e800, 0xf, 0x11, 0xc0001aa4e0, 0x4, 0x3, 0xc0012f2b00)
	/home/rspazzol/go/src/github.com/redhat-cop/egressip-ipam-operator/pkg/controller/egressipam/ipam.go:283 +0x289b
github.com/redhat-cop/egressip-ipam-operator/pkg/controller/egressipam.(*ReconcileEgressIPAM).Reconcile(0xc000982900, 0x0, 0x0, 0xc0024218f0, 0xe, 0xc000812cd8, 0xc0004e6000, 0xc0004373b8, 0x1d64e00)
	/home/rspazzol/go/src/github.com/redhat-cop/egressip-ipam-operator/pkg/controller/egressipam/egressipam_controller.go:382 +0xf57
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000546540, 0x18ab9e0, 0xc001159e00, 0x440700)
	/home/rspazzol/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:256 +0x162
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000546540, 0x0)
	/home/rspazzol/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:232 +0xcb
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc000546540)
	/home/rspazzol/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:211 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc00098a1d0)
	/home/rspazzol/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00098a1d0, 0x3b9aca00, 0x0, 0x1, 0xc000a7d380)
	/home/rspazzol/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc00098a1d0, 0x3b9aca00, 0xc000a7d380)
	/home/rspazzol/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/home/rspazzol/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:193 +0x328
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0

goroutine 1712 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/home/rspazzol/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/runtime/runtime.go:55 +0x105
panic(0x197a120, 0xc00234aa40)
	/usr/lib/golang/src/runtime/panic.go:679 +0x1b2
github.com/redhat-cop/egressip-ipam-operator/pkg/controller/egressipam.(*ReconcileEgressIPAM).assignIPsToNodes(0xc000982900, 0xc0015352c0, 0xc00013e800, 0xf, 0x11, 0xc0001aa4e0, 0x4, 0x3, 0xc0012f2b00)
	/home/rspazzol/go/src/github.com/redhat-cop/egressip-ipam-operator/pkg/controller/egressipam/ipam.go:283 +0x289b
github.com/redhat-cop/egressip-ipam-operator/pkg/controller/egressipam.(*ReconcileEgressIPAM).Reconcile(0xc000982900, 0x0, 0x0, 0xc0024218f0, 0xe, 0xc000812cd8, 0xc0004e6000, 0xc0004373b8, 0x1d64e00)
	/home/rspazzol/go/src/github.com/redhat-cop/egressip-ipam-operator/pkg/controller/egressipam/egressipam_controller.go:382 +0xf57
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000546540, 0x18ab9e0, 0xc001159e00, 0x440700)
	/home/rspazzol/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:256 +0x162
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000546540, 0x0)
	/home/rspazzol/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:232 +0xcb
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc000546540)
	/home/rspazzol/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:211 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc00098a1d0)
	/home/rspazzol/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00098a1d0, 0x3b9aca00, 0x0, 0x1, 0xc000a7d380)
	/home/rspazzol/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc00098a1d0, 0x3b9aca00, 0xc000a7d380)
	/home/rspazzol/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/home/rspazzol/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:193 +0x328
```

--- a/pkg/controller/egressipam/hostsubnet.go
+++ b/pkg/controller/egressipam/hostsubnet.go
@@ -6,6 +6,7 @@ import (
 
 	ocpnetv1 "github.com/openshift/api/network/v1"
 	redhatcopv1alpha1 "github.com/redhat-cop/egressip-ipam-operator/pkg/apis/redhatcop/v1alpha1"
+	"github.com/scylladb/go-set/strset"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/client-go/util/workqueue"
@@ -94,7 +95,7 @@ func (r *ReconcileEgressIPAM) reconcileHSAssignedIPs(nodeAssignedIPs map[string]
 			log.Error(err, "unable to lookup hostsubnet for ", "node", node.GetName)
 			return err
 		}
-		if !reflect.DeepEqual(nodeAssignedIPs[node.GetName()], hostsubnet.EgressIPs) {
+		if !strset.New(nodeAssignedIPs[node.GetName()]...).IsEqual(strset.New(hostsubnet.EgressIPs...)) {
 			hostsubnet.EgressIPs = nodeAssignedIPs[node.GetName()]
 			err := r.GetClient().Update(context.TODO(), &hostsubnet, &client.UpdateOptions{})
 			if err != nil {
@@ -127,7 +128,7 @@ func (r *ReconcileEgressIPAM) assignCIDRsToHostSubnets(nodesByCIDR map[string][]
 				log.Error(err, "unable to lookup hostsubnet for ", "node", node)
 				return err
 			}
-			if !reflect.DeepEqual(hostsubnet.EgressCIDRs, []string{cidrByNode[node.GetName()]}) {
+			if !strset.New(hostsubnet.EgressCIDRs...).IsEqual(strset.New([]string{cidrByNode[node.GetName()]}...)) {
 				hostsubnet.EgressCIDRs = []string{cidrByNode[node.GetName()]}
 				err := r.GetClient().Update(context.TODO(), &hostsubnet, &client.UpdateOptions{})
 				if err != nil {

--- a/pkg/controller/egressipam/ipam.go
+++ b/pkg/controller/egressipam/ipam.go
@@ -91,14 +91,14 @@ func (r *ReconcileEgressIPAM) assignIPsToNamespaces(unassignedNamespaces []corev
 	}
 	log.V(1).Info("sorted reserved IPs ", "IPs by CIDR", IPsByCIDR)
 	for i := range unassignedNamespaces {
-		IPs, err := getNextAvailableIPs(IPsByCIDR)
+		newIPsByCIDRs, err := getNextAvailableIPs(IPsByCIDR)
 		if err != nil {
 			log.Error(err, "unable to assing new IPs for ", "namespace", unassignedNamespaces[i])
 			return []corev1.Namespace{}, err
 		}
 		ipstrings := []string{}
-		for _, IP := range IPs {
-			ipstrings = append(ipstrings, IP.String())
+		for _, cidrAssignment := range egressIPAM.Spec.CIDRAssignments {
+			ipstrings = append(ipstrings, newIPsByCIDRs[cidrAssignment.CIDR].String())
 		}
 		log.Info("ips assigned to", "namespace", unassignedNamespaces[i].GetName(), "ips", ipstrings)
 		unassignedNamespaces[i].ObjectMeta.Annotations[namespaceAssociationAnnotation] = strings.Join(ipstrings, ",")
@@ -113,18 +113,18 @@ func (r *ReconcileEgressIPAM) assignIPsToNamespaces(unassignedNamespaces []corev
 
 // returns a set of IPs. These IPs are the next available IP per CIDR.
 // The map of CIDR is passed by reference and updated with the new IPs, so this function can be used in a loop.
-func getNextAvailableIPs(IPsByCIDR map[string][]net.IP) ([]net.IP, error) {
+func getNextAvailableIPs(IPsByCIDR map[string][]net.IP) (map[string]net.IP, error) {
 	log.V(1).Info("Assigning new IPs from", "IPs by CIDR", IPsByCIDR)
-	iPsByCIDR := IPsByCIDR
-	assignedIPs := []net.IP{}
-	for cidr := range iPsByCIDR {
-		assignedIP, newIPs, err := getNextAvailableIP(cidr, iPsByCIDR[cidr])
+	//iPsByCIDR := IPsByCIDR
+	assignedIPs := map[string]net.IP{}
+	for cidr := range IPsByCIDR {
+		assignedIP, newIPs, err := getNextAvailableIP(cidr, IPsByCIDR[cidr])
 		if err != nil {
 			log.Error(err, "unable to assign get next ip for", "cidr", cidr)
-			return []net.IP{}, err
+			return map[string]net.IP{}, err
 		}
-		assignedIPs = append(assignedIPs, assignedIP)
-		iPsByCIDR[cidr] = newIPs
+		assignedIPs[cidr] = assignedIP
+		IPsByCIDR[cidr] = newIPs
 	}
 	log.V(1).Info("Assigned", "new IPs from", assignedIPs, " new IPs by CIDR", IPsByCIDR)
 	return assignedIPs, nil
@@ -185,12 +185,10 @@ func (r *ReconcileEgressIPAM) assignIPsToNodes(assignedIPsByNode map[string][]st
 	assignedIPsToNodesByCIDR := map[string][]string{}
 	assignedIPsToNamespaceByCIDR := map[string][]string{}
 	toBeAssignedToNodesIPsByCIDR := map[string][]string{}
-	toBeRemovedIPsFromNodesByCIDR := map[string][]string{}
 	for _, cidrAssignment := range egressIPAM.Spec.CIDRAssignments {
 		assignedIPsToNodesByCIDR[cidrAssignment.CIDR] = []string{}
 		assignedIPsToNamespaceByCIDR[cidrAssignment.CIDR] = []string{}
 		toBeAssignedToNodesIPsByCIDR[cidrAssignment.CIDR] = []string{}
-		toBeRemovedIPsFromNodesByCIDR[cidrAssignment.CIDR] = []string{}
 	}
 	// 1. get assignedIPsToNodesByCIDR
 	for _, ipsbn := range assignedIPsByNode {
@@ -266,23 +264,37 @@ func (r *ReconcileEgressIPAM) assignIPsToNodes(assignedIPsByNode map[string][]st
 		return map[string][]string{}, err
 	}
 
+	//log.V(1).Info("", "nodesByCIDR: ", nodesByCIDR)
+
 	// 7. calculate NodesByNumberOfAssignedIPByCIDR
-	nodesByNumberOfAssignedIPsByCIDR := map[string]map[int][]corev1.Node{}
+	nodesByNumberOfAssignedIPsByCIDR := map[string]map[int][]string{}
 	for cidr := range toBeAssignedToNodesIPsByCIDR {
-		nodesByNumberOfAssignedIPsByCIDR[cidr] = map[int][]corev1.Node{}
+		nodesByNumberOfAssignedIPsByCIDR[cidr] = map[int][]string{}
+		nodes := []string{}
 		for _, node := range nodesByCIDR[cidr] {
-			nodesByNumberOfAssignedIPsByCIDR[cidr][len(assignedIPsToNodesByCIDR[cidr])] = append(nodesByNumberOfAssignedIPsByCIDR[cidr][len(assignedIPsToNodesByCIDR[cidr])], node)
+			nodes = append(nodes, node.GetName())
 		}
+		nodesByNumberOfAssignedIPsByCIDR[cidr][len(nodes)] = nodes
 	}
+
+	log.V(1).Info("", "nodesByNumberOfAssignedIPsByCIDR: ", nodesByNumberOfAssignedIPsByCIDR)
 
 	// 8. assign IPs to the least assigned nodes, update map, by CIDR
 	for cidr, ips := range toBeAssignedToNodesIPsByCIDR {
 		for _, ip := range ips {
 			//pick the first node with the least IPs in this CIDR
+			log.V(1).Info("", "nodesByNumberOfAssignedIPsByCIDR: ", nodesByNumberOfAssignedIPsByCIDR)
 			minIPsPerNode := getMinKey(nodesByNumberOfAssignedIPsByCIDR[cidr])
+			if minIPsPerNode == -1 {
+				err := errors.New("Unable to find nodes for CIDR" + cidr)
+				log.Error(err, "", cidr, "nodes", nodesByNumberOfAssignedIPsByCIDR[cidr])
+				return map[string][]string{}, err
+			}
+			log.V(1).Info("", "minIPsPerNode: ", minIPsPerNode, "for cidr", cidr)
 			node := nodesByNumberOfAssignedIPsByCIDR[cidr][minIPsPerNode][0]
+			log.Info("assigning", "IP", ip, "to node", node)
 			// add the node to the assignedIP per node map
-			assignedIPsByNode[node.GetName()] = append(assignedIPsByNode[node.GetName()], ip)
+			assignedIPsByNode[node] = append(assignedIPsByNode[node], ip)
 			// remove the node from the minIPsPerNode map
 			nodesByNumberOfAssignedIPsByCIDR[cidr][minIPsPerNode] = nodesByNumberOfAssignedIPsByCIDR[cidr][minIPsPerNode][1:]
 			// add the node to the minIPsPerNode+1 map
@@ -293,10 +305,15 @@ func (r *ReconcileEgressIPAM) assignIPsToNodes(assignedIPsByNode map[string][]st
 	return assignedIPsByNode, nil
 }
 
-func getMinKey(nodemap map[int][]corev1.Node) int {
+func getMinKey(nodemap map[int][]string) int {
 	numbers := []int{}
-	for n := range nodemap {
-		numbers = append(numbers, n)
+	for n, nodes := range nodemap {
+		if len(nodes) > 0 {
+			numbers = append(numbers, n)
+		}
+	}
+	if len(numbers) == 0 {
+		return -1
 	}
 	sort.Ints(numbers)
 	return numbers[0]

--- a/readme.md
+++ b/readme.md
@@ -140,7 +140,7 @@ oc apply -f deploy/role.yaml -n egressip-ipam-operator
 oc apply -f deploy/role_binding.yaml -n egressip-ipam-operator
 export token=$(oc serviceaccounts get-token 'egressip-ipam-operator' -n egressip-ipam-operator)
 oc login --token=${token}
-OPERATOR_NAME='egressip-ipam-operator' NAMESPACE='egressip-ipam-operator' operator-sdk --verbose up local --namespace "" --operator-flags="--zap-level=debug"
+OPERATOR_NAME='egressip-ipam-operator' NAMESPACE='egressip-ipam-operator' operator-sdk --verbose run --local --namespace "" --operator-flags="--zap-level=debug"
 ```
 
 ## Testing

--- a/test/test.sh
+++ b/test/test.sh
@@ -1,0 +1,25 @@
+#/bin/bash
+COUNT=1
+while [ $COUNT -lt 40 ]; do
+
+echo "
+
+apiVersion: v1
+
+kind: Namespace
+
+metadata:
+
+  annotations:
+
+    egressip-ipam-operator.redhat-cop.io/egressipam: egressipam-aws
+
+  name: ics-egressip-test${COUNT}
+
+" | oc apply -f -
+
+COUNT=`expr $COUNT + 1`
+
+sleep 10
+
+done

--- a/test/untest.sh
+++ b/test/untest.sh
@@ -1,0 +1,8 @@
+COUNT=1
+while [ $COUNT -lt 40 ]; do
+
+oc delete namespace ics-egressip-test${COUNT}
+
+COUNT=`expr $COUNT + 1`
+
+done

